UserScrape/crawler.py

```diff
@@ -121,108 +135,145 @@ def login(self) -> CrawlResult:
-                                     'channel': self.wait.until(EC.presence_of_element_located(
-                                         (By.CSS_SELECTOR,
-                                          "ytd-channel-name.ytd-video-owner-renderer > div:nth-child(1) > "
-                                          "div:nth-child(1)"))).text,
-                                     'channel_id': self.wait.until(EC.presence_of_element_located(
-                                         (By.CSS_SELECTOR, "#text > a"))).get_attribute('href').strip(
-                                         'https://www.youtube.com/channel/')
-                                     }
-
-    def get_recommendations_for_video(self, source, branching):
+    def get_recommendations_for_video(self, source):
         self.driver.get("https://www.youtube.com/watch?v=" + source)
-        results_content = WebDriverWait(self.driver, 10).until(
-            EC.visibility_of_element_located((By.XPATH, '// *[ @ id = "upnext"]')))
-        all_recs = WebDriverWait(self.driver, 10).until(
-            EC.presence_of_all_elements_located((By.XPATH, '//*[@id="dismissable"]/div/div[1]/a'))
-        )
 
-        recos = []
+        # this is the list of elements from the recommendation sidebar
+        # it does not always load all recommendations at the same time, therefore the loop
+        all_recs = []
+        while len(all_recs) < 19:
+            all_recs = WebDriverWait(self.driver, 10).until(
+                EC.visibility_of_all_elements_located(
+                    (By.XPATH, '//*[@id="dismissable"]/div/div[1]/a'))
+            )
+
+        recos = {}
+        personalized_counter = 0 # how many of the recommendations are personalized?
         for i in all_recs:
-            # somehow there are also empty elements with the id video-title found, these need to be removed
-            if len(i.text) != 0:
-                recos.append(i.get_attribute('href').replace('https://www.youtube.com/watch?v=', ''))
-        self.get_video_features(source, recos)
-        return recos[0:branching]
+            personalized = 'Recommended for you' in i.text
+            if personalized:
+                personalized_counter += 1
+            # take the link and remove everything except for the id of the video that the link leads to
+            recommendation_id = i.get_attribute('href').replace(
+                'https://www.youtube.com/watch?v=', '')
+            recos[recommendation_id] = {
+                'id': recommendation_id, 'personalized': personalized}
+        # store the information about the current video plus the corresponding recommendations
+        self.get_video_features(source, recos, personalized_counter)
+        # return the recommendations 
+        return recos
+
+    def delete_last_video_from_history(self):
+        self.driver.get('https://www.youtube.com/feed/history')
+        # todo: maybe replace the full xpath because its ugly, but it works
```

Comment:
If possible I would like to replace this extremely long xpath, if you have a good idea let me know

Reply:
its going to be pretty brittle. Can we execute a JS script and get the results in selenium?
this is the best way to find it I think:
```javascript
Array.from(document.querySelectorAll('ytd-video-renderer ytd-menu-renderer yt-icon-button#button')).filter(n => n.__data.label.startsWith("Remove"))
```

Reply:
replaced by a shorter but unfortunately language dependent xpath. I used it in German because somehow the interface was in german if I ran it in headless mode. I switched to english now. This will probably not be a problem as soon as it runs in production since we use english accounts with us proxys (right?)